gocr-0.50/man/man1/gocr.1
.\"             -*-Nroff-*-
.\" Copyright 2000 Red Hat, Inc.
.TH GOCR 1 "29 Mar 2009" "Linux" "User's Manual"
.SH NAME
gocr \- command line text recognition tool
.SH SYNOPSIS
.B gocr
[\fIOPTION\fR] [\fB-i\fR] \fIpnm-file\fR
.fi
.SH DESCRIPTION
gocr is an optical character recognition program that can be used from
the command line.  It takes input in PNM, PGM, PBM, PPM, or PCX
format, and writes recognized text to \fBstdout\fR.  If the \fIpnm
file\fR is a single dash, PNM data is read from \fBstdin\fR.
If gzip, bzip2 and netpbm-progs are installed and your system
supports popen(3) also pnm.gz, pnm.bz2, png, jpg, jpeg, tiff,
gif, bmp, ps (only single pages) and eps are supported as input files
(not as input stream), where pnm can be replaced
by one of ppm, pgm and pbm.
.SH OPTIONS
.TP
\fB\-h\fR
show usage information
.TP
\fB\-i\fR \fIfile\fR
read input from \fIfile\fR (or \fBstdin\fR if \fIfile\fR is a single
dash)
.TP
\fB\-o\fR \fIfile\fR
send output to \fIfile\fR instead of \fBstdout\fR
.TP
\fB\-e\fR \fIfile\fR
send errors to \fIfile\fR instead of \fBstderr\fR or to \fBstdout\fR
if \fIfile\fR is a dash
.TP
\fB\-x\fR \fIfile\fR
progress output to \fIfile\fR
(\fIfile\fR can be a file name, a fifo name or a file descriptor 1...255),
this is useful for GUI developpers to show the OCR progress, the file
descriptor argument is only available, if compiled with __USE_POSIX defined
.TP
\fB\-p\fR \fIpath\fR
database path, a final slash must be included, default is ./db/,
this path will be populated with images of learned characters
.TP
\fB\-f\fR \fIformat\fR
output \fIformat\fR of the recognized text 
(ISO8859_1 TeX HTML XML UTF8 ASCII), XML will also output position and
probability data 
.TP
\fB\-l\fR \fIlevel\fR
set grey level to \fIlevel\fR (0<160<=255, default:
0 for autodetect), darker pixels belong to characters, brighter pixels
are interpreted as background of the input image
.TP
\fB\-d\fR \fIsize\fR
set dust size in pixels (clusters smaller than this are removed),
0 means no clusters are removed, the default is -1 for auto detection
.TP
\fB\-s\fR \fInum\fR
set spacewidth between words in units of dots (default: 0 for autodetect),
wider widths are interpreted as word spaces, smaller as character spaces
.TP
\fB\-v\fR \fIverbosity\fR
be verbose to stderr; \fIverbosity\fR is a bitfield
.TP
\fB\-c \fIstring\fR
only verbose output of characters from \fIstring\fR to stderr,
more output is generated for all characters within the string,
the underscore stands for unknown chars, this function is usefull
to limit debug information to the necessary one 
.TP
\fB\-C \fIstring\fR
only recognise characters from \fIstring\fR, this is a filter function
in cases where the interest is only to a part of the character alphabet,
you can use 0-9 or a-z to specify ranges, use -- to detect the minus sign
.TP
\fB\-a\fR \fIcertainty\fR
set value for certainty of recognition (0..100; default: 95),
characters with a higher certainty are accepted, characters with a lower
certainty are treated as unknown (not recognized); set higher values, if
you want to have only more certain recognized characters
.TP
\fB\-u \fIstring\fR
output this string for every unrecognized character (default is "_")
.TP
\fB\-m\fR \fImode\fR
set oprational mode; mode is a bitfield (default: 0)
.TP
\fB\-n\fR \fIbool\fR
if \fIbool\fR is non-zero, only recognise numbers
(this is now obsolete, use -C "0123456789")
.PP
The verbosity is specified as a bitfield:
.TP 10
.B 1
print more info
.TP
.B 2
list shapes of boxes (see -c) to stderr
.TP
.B 4
list pattern of boxes (see -c) to stderr
.TP
.B 8
print pattern after recognition for debugging
.TP
.B 16
print debug information about recognition of lines to stderr
.TP
.B 32
create outXX.png with boxes and lines marked on each general OCR-step
.PP
The operation modes are:
.TP 10
.B 2
use database to recognize characters which are not recognized by
other algorithms, (early development)
.TP
.B 4
switching on layout analysis or zoning (development)
.TP
.B 8
don't compare unrecognized characters to recognized one
.TP
.B 16
don't try to divide overlapping characters to two or three single characters 
.TP
.B 32
don't do context correction
.TP
.B 64
character packing, before recognition starts, similar characters
are searched and only one of this characters will be send to the
recognition engine (development)
.TP
.B 130
extend database, prompts user for unidentified characters
and extends the database with users answer (128+2, early development)
.TP
.B 256
switch off the recognition engine (makes sense together with -m 2)
.SH
.SH AUTHOR
Joerg Schulenburg (see http://jocr.sourceforge.net/ for EMAIL)
.br
First version of man page by Tim Waugh <twaugh@redhat.com>
.SH VERSION INFORMATION
This man page documents \fBgocr\fR, version 0.41.
.SH REPORTING BUGS
Report bugs to Joerg Schulenburg
.SH SEE ALSO
More details can be found at /usr/share/doc/gocr-\fBX.XX\fR/gocr.html.
Also read /usr/share/doc/gocr-\fBX.XX\fR/README to learn, how to improve
results.
.SH EXAMPLES
.TP
.B gocr -v 33 text1.pbm
output verbose information, out30.png is created to see details of
recognition process
.TP
.B gocr -v 7 -c _YV text1.pbm
verbose output for unknown chars and chars Y and V
.TP
.B djpeg -pnm -gray text.jpg | gocr -
convert a jpeg file to pnm format and input via pipe


gocr-0.50/HISTORY
History: (Changes,ChangeLog)

 0.51 in progress
   
 0.50 Sep10-May12,Mar13
   just release it to avoid questiions to old problems, give a life sign ;)
   fix 4 parfait problems against 0.48 (thanks to Rich Burridge) 
   adding qrcode detection and decoding (no error correction, no skewing)
   spacing slightly improved
   context correction of hex codes (p.e. hex fingerprints)
   some threshold value adaptions (not finished)
   try to fix double output of XML code <...> and removed additional \n
   improved quotation detection ,, ''
   improved monospaced spacing (video text)

 0.49 Aug09-Sep10
   fix dot handling for ':' and ';' (vector code)
   fix '@' for 7x9 and 5x8 fonts
   fix double counting of subboxes (affects "0" (zero) with dot in it)
   character "l" of width 1 improved
   bug fix gluing chars ij of width=1
   bug fix thresholding (small gray images)
   return error code -1 on ERROR pnm.c unexpected EOF
   fix conflicts with unicode_defs.h TRUE definition on gcc/alphaev7-osf/3.4
   further fixes for lib by D. Katsubo
   fix #3039007 "struct list" in list.h conflicts with STL (ocr_object_list)
   fix #3039006 INFINITY macro in unicode.h conflicts with math.h
   bugfix barcode 128, switch from mode mC to mA (":1")
   bugfix: MultiPNM + database  - ID: 2957140
   improved barcode recognition - ID: 2859644 (bars wider than spaces)
   quality test-script bin/gocr_chk.sh added
   initial datamatrix support (ASCII + ASCII numeric only, no ErrCorrection)

 0.48 Jul09
       fix buffer overflow introduced in 0.46 for filenames
       add codabar barcode
       fix bug, removing melted serifs
       add patch by Chris Lee, i25 barcode recognition + modifications
       fix some false positive numbers "34" (video, gas meter)
       fix problems with 2zZ4 for 10x10 screen font
       better debug output for :;,.
       remove examples, doc and libs part from configure (see below)
       remove doc and examples from the (make install) part to reduce
         dependencies (gs and transfig is not needed for rpm/ebuild)
         gocr only may depend from netpbm, but can live without too
         this will help to install gocr on "exotic" (nonlinux) platforms
       fix gentoo app-text/gocr Bug 243250 src/Makefile: $(CC) $(LDFLAGS) ...

 0.47  fix database recognition for certainty 100 (-a 100)
       insert spaces with certainty 100 (old: 99) to let -a 100 work
       new option -u string for unrecognized chars
       fix: No contrast in image causes division by zero
       reduced false positive recognition of scanned "a496" (Gutenberg Project)
       "d as a" patch ID: 1556112
       add "Windows Pipe Fix", but I hate extra code for bad environments
       improve 7x10, sample 0811qemu1.png (ToDo: not finished)
       change black:white from &gt;4:1 to &gt;3.5:1 as criteria of inversion
       reintroduce static library libPgm2asc.a (make libs) for OSRA project
       add dynamic library (make libs), unused but may help other projects
      
 0.46  improved context correction (especially helvetica "Il")
       improved recognition of tiny chars "$1", fat "s", "rw" ","
       fix blank spaces problem in filenames
         (http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=316511)
         !!! please check on other platforms and report to me !!!
         there are still problems with special chars (double quodes, backslash)
         better use this way: djpeg -gray -pnm strangefilename.jpg | gocr -
       fix possible problem with database and UTF8 input 
       fix hidden bug in pitch/spacing initialization
       reactivate code for output of glued chars and strings 
       fix wrong close() call
       remove creation of pgm2asc.a for simplicity (see SF-patch 1827477)
       
 0.45  minor corrections for c and k
       minus sign is filtered by option -C "--" now, ("\-" was parsed badly)
       clean up old unused code for simplicity (api, frontend)
       fix problem with low height barcodes and barcode removing
       fix problem with readpgm (for multiple images) and database
       PACKAGE_VERSION defined by configure.in AC_INIT + gocr.spec

 0.44  add volume to boxes (negative means white areas inside black areas)
       Fix overflow in despeckling routine (verbose mode, dust removing)
       reactivate composed chars, fix merge_boxes
       fix problems with uncertain line detection and not recognized "7"
       option -a has an effect now for the output
       adaptions to MICR E13-B font (see GnuMICR), ToDo: 4 extra-chars
       fix num_boxes in merge_boxes (affects line detection)
       reduce 2 prompts to one per char in database mode, ^A for skip all
       fix problem with smaller headlines
       fix problems with tall font (4)
       fix includes for non-linux-platforms

 0.43  fix problem with dark frame around image
       support multiple images, ex: giftopnm -image=all a.gif | gocr -
       invert if obviously white on black (black_mass>=4*white_mass)
       improve thresholding for discrete histograms
         (note: this can particularly lead to bad results, will be fixed later)
       speedup for big boxes (especially dark background)
       fix memory leak (setas(same string) + detect_barcode)
       fix uninitialized variables after insert spaces (num_frames)
       fix frame_vector for single pixels (twice + ERROR idx out of range)

 0.42  further parts of recognition engine relaced by vector version
       changed colored debug output for out??.png
       division of glued chars replaced (slower but more accurate)
       fix framing of small font
       fix problem with uninitialized pnm_readpaminit call (CPS 21Nov06)
       better progress output (see progress.[ch]), new image debug output
       switch to the new improved rotation detection

 0.41  (buggy if --with-netpbm=no, apply the pgm-patch!)
       otsu.c concentrates now only on high contrast regions
       fix pnm reads for 2 byte pixels (--with-libpbm=no) 

gocr-0.50/src/gocr.c
/*
This is a Optical-Character-Recognition program
Copyright (C) GPLv2 2000-2013  Joerg Schulenburg

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

 see README for EMAIL-address

  sometimes I have written comments in german language, sorry for that

 This file was retrieved from pgm2asc.cc of Joerg, in order to have
 a library of the ocr-engine from Klaas Freitag
 
 */
#include "config.h"
#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include <string.h>
#ifdef HAVE_GETTIMEOFDAY 
#include <sys/time.h>
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#include "pnm.h"
#include "pgm2asc.h"
#include "pcx.h"
#include "ocr0.h"		/* only_numbers */
#include "progress.h"
#include "version.h"

static void out_version(int v) {
  fprintf(stderr, " Optical Character Recognition --- gocr "
          version_string " " release_string "\n"
          " Copyright (C) 2001-2010 Joerg Schulenburg  GPG=1024D/53BDFBE3\n"
          " released under the GNU General Public License\n");
  /* as recommended, (c) and license should be part of the binary */
  /* no email because of SPAM, see README for contacting the author */
  if (v)
    fprintf(stderr, " use option -h for help\n");
  if (v & 2)
    exit(1);
  return;
}

static void help(void) {
  out_version(0);
  /* output is shortened to essentials, see manual page for details */
  fprintf(stderr,
	  " using: gocr [options] pnm_file_name  # use - for stdin\n"
	  " options (see gocr manual pages for more details):\n"
	  " -h, --help\n"
	  " -i name   - input image file (pnm,pgm,pbm,ppm,pcx,...)\n"
	  " -o name   - output file  (redirection of stdout)\n"
	  " -e name   - logging file (redirection of stderr)\n"
	  " -x name   - progress output to fifo (see manual)\n"
	  " -p name   - database path including final slash (default is ./db/)\n");
  fprintf(stderr, /* string length less than 509 bytes for ISO C89 */
	  " -f fmt    - output format (ISO8859_1 TeX HTML XML UTF8 ASCII)\n"
	  " -l num    - threshold grey level 0<160<=255 (0 = autodetect)\n"
	  " -d num    - dust_size (remove small clusters, -1 = autodetect)\n"
	  " -s num    - spacewidth/dots (0 = autodetect)\n"
	  " -v num    - verbose (see manual page)\n"
	  " -c string - list of chars (debugging, see manual)\n"
	  " -C string - char filter (ex. hexdigits: ""0-9A-Fx"", only ASCII)\n"
	  " -m num    - operation modes (bitpattern, see manual)\n");
  fprintf(stderr, /* string length less than 509 bytes for ISO C89 */
	  " -a num    - value of certainty (in percent, 0..100, default=95)\n"
	  " -u string - output this string for every unrecognized character\n");
  fprintf(stderr, /* string length less than 509 bytes for ISO C89 */
	  " examples:\n"
	  "\tgocr -m 4 text1.pbm                   # do layout analyzis\n"
	  "\tgocr -m 130 -p ./database/ text1.pbm  # extend database\n"
	  "\tdjpeg -pnm -gray text.jpg | gocr -    # use jpeg-file via pipe\n"
	  "\n");
  fprintf(stderr, " webpage: http://jocr.sourceforge.net/ (may out of date)\n");
  fprintf(stderr, " mirror:  http://www-e.uni-magdeburg.de/jschulen/ocr/\n");
  exit(0);
}

#ifdef HAVE_GETTIMEOFDAY
/* from the glibc documentation */
static int timeval_subtract (struct timeval *result, struct timeval *x, 
    struct timeval *y) {

  /* Perform the carry for the later subtraction by updating Y. */
  if (x->tv_usec < y->tv_usec) {
    int nsec = (y->tv_usec - x->tv_usec) / 1000000 + 1;
    y->tv_usec -= 1000000 * nsec;
    y->tv_sec += nsec;
  }
  if (x->tv_usec - y->tv_usec > 1000000) {
    int nsec = (x->tv_usec - y->tv_usec) / 1000000;
    y->tv_usec += 1000000 * nsec;
    y->tv_sec -= nsec;
  }

  /* Compute the time remaining to wait.
     `tv_usec' is certainly positive. */
  result->tv_sec = x->tv_sec - y->tv_sec;
  result->tv_usec = x->tv_usec - y->tv_usec;

  /* Return 1 if result is negative. */
  return x->tv_sec < y->tv_sec;
}
#endif

static void process_arguments(job_t *job, int argn, char *argv[])
{
  int i;
  char *s1;

  assert(job);

  if (argn <= 1) {
    out_version(1);
    exit(0);
  }
#ifdef HAVE_PGM_H
  pnm_init(&argn, &argv);
#endif
  
  /* process arguments */
  for (i = 1; i < argn; i++) {
    if (strcmp(argv[i], "--help") == 0)
      help(); /* and quits */
    if (argv[i][0] == '-' && argv[i][1] != 0) {
      s1 = "";
      if (i + 1 < argn)
	s1 = argv[i + 1];
      switch (argv[i][1]) {
      case 'h': /* help */
	help();
	break;
      case 'i': /* input image file */
	job->src.fname = s1;
	i++;
	break;
      case 'e': /* logging file */
	if (s1[0] == '-' && s1[1] == '\0') {
#ifdef HAVE_UNISTD_H
          dup2(STDOUT_FILENO, STDERR_FILENO); /* -e /dev/stdout  works */
#else
	  fprintf(stderr, "stderr redirection not possible without unistd.h\n");
#endif           
	}
	else if (!freopen(s1, "w", stderr)) {
	  fprintf(stderr, "stderr redirection to %s failed\n", s1);
	}
	i++;
	break;
      case 'p': /* database path */
	job->cfg.db_path=s1;
	i++;
	break;
      case 'o': /* output file */
	if (s1[0] == '-' && s1[1] == '\0') {	/* default */
	}
	else if (!freopen(s1, "w", stdout)) {
	  fprintf(stderr, "stdout redirection to %s failed\n", s1);
	};
	i++;
	break;
      case 'f': /* output format */
        if (strcmp(s1, "ISO8859_1") == 0) job->cfg.out_format=ISO8859_1; else
        if (strcmp(s1, "TeX")       == 0) job->cfg.out_format=TeX; else
        if (strcmp(s1, "HTML")      == 0) job->cfg.out_format=HTML; else
        if (strcmp(s1, "XML")       == 0) job->cfg.out_format=XML; else
        if (strcmp(s1, "SGML")      == 0) job->cfg.out_format=SGML; else
        if (strcmp(s1, "UTF8")      == 0) job->cfg.out_format=UTF8; else
        if (strcmp(s1, "ASCII")     == 0) job->cfg.out_format=ASCII; else
        fprintf(stderr,"Warning: unknown format (-f %s)\n",s1);
        i++;
        break;
      case 'c': /* list of chars (_ = not recognized chars) */
	job->cfg.lc = s1;
	i++;
	break;
      case 'C': /* char filter, default: NULL (all chars) */
        /* ToDo: UTF8 input, wchar */
	job->cfg.cfilter = s1;
	i++;
	break;
      case 'd': /* dust size */
	job->cfg.dust_size = atoi(s1);
	i++;
	break;
      case 'l': /* grey level 0<160<=255, 0 for autodetect */
	job->cfg.cs = atoi(s1);
	i++;
	break;
      case 's': /* spacewidth/dots (0 = autodetect) */
	job->cfg.spc = atoi(s1);
	i++;
	break;
      case 'v': /* verbose mode */
	job->cfg.verbose |= atoi(s1);
	i++;
	break;
      case 'm': /* operation modes */
	job->cfg.mode |= atoi(s1);
	i++;
	break;
      case 'n': /* numbers only */
	job->cfg.only_numbers = atoi(s1);
	i++;
	break;
      case 'x': /* initialize progress output s1=fname */
	ini_progress(s1);
	i++;
	break;
      case 'a': /* set certainty */
	job->cfg.certainty = atoi(s1);;
	i++;
	break;
      case 'u': /* output marker for unrecognized chars */
        job->cfg.unrec_marker = s1;
        i++;
        break;
      default:
	fprintf(stderr, "# unknown option use -h for help\n");
      }
      continue;
    }
    else /* argument can be filename v0.2.5 */ if (argv[i][0] != '-'
						   || argv[i][1] == '\0' ) {
      job->src.fname = argv[i];
    }
  }
}

static void mark_start(job_t *job) {
  assert(job);

  if (job->cfg.verbose) {
    out_version(0);
    /* insert some helpful info for support */
    fprintf(stderr, "# compiled: " __DATE__ );
#if defined(__GNUC__)
    fprintf(stderr, " GNUC-%d", __GNUC__ );
#endif
#ifdef __GNUC_MINOR__
    fprintf(stderr, ".%d", __GNUC_MINOR__ );
#endif
#if defined(__linux)
    fprintf(stderr, " linux");
#elif defined(__unix)
    fprintf(stderr, " unix");
#endif
#if defined(__WIN32) || defined(__WIN32__)
    fprintf(stderr, " WIN32");
#endif
#if defined(__WIN64) || defined(__WIN64__)
    fprintf(stderr, " WIN64");
#endif
#if defined(__VERSION__)
    fprintf(stderr, " version " __VERSION__ );
#endif
    fprintf(stderr, "\n");
    fprintf(stderr,
      "# options are: -l %d -s %d -v %d -c %s -m %d -d %d -n %d -a %d -C \"%s\"\n",
      job->cfg.cs, job->cfg.spc, job->cfg.verbose, job->cfg.lc, job->cfg.mode, 
      job->cfg.dust_size, job->cfg.only_numbers, job->cfg.certainty,
      job->cfg.cfilter);
    fprintf(stderr, "# file: %s\n", job->src.fname);
#ifdef USE_UNICODE
    fprintf(stderr,"# using unicode\n");
#endif
#ifdef HAVE_GETTIMEOFDAY
    gettimeofday(&job->tmp.init_time, NULL);
#endif
  }
}

static void mark_end(job_t *job) {
  assert(job);

#ifdef HAVE_GETTIMEOFDAY
  /* show elapsed time */
  if (job->cfg.verbose) {
    struct timeval end, result;
    gettimeofday(&end, NULL);
    timeval_subtract(&result, &end, &job->tmp.init_time);
    fprintf(stderr,"Elapsed time: %d:%02d:%3.3f.\n", (int)result.tv_sec/60,
	(int)result.tv_sec%60, (float)result.tv_usec/1000);
  }
#endif
}

static int read_picture(job_t *job) {
  int rc=0;
  assert(job);

  if (strstr(job->src.fname, ".pcx"))
    readpcx(job->src.fname, &job->src.p, job->cfg.verbose);
  else
    rc=readpgm(job->src.fname, &job->src.p, job->cfg.verbose);
  return rc; /* 1 for multiple images, -1 on error, 0 else */
}

/* subject of change, we need more output for XML (ToDo) */
void print_output(job_t *job) {
  int linecounter = 0;
  const char *line;

  assert(job);
        
/* TODO: replace getTextLine-loop(line) by output_text() (libs have to use pipes)
   simplify code 2010-09-26
*/
  linecounter = 0;
  line = getTextLine(&(job->res.linelist), linecounter++);
  while (line) {
    /* notice: decode() is shiftet to getTextLine since 0.38 */
    fputs(line, stdout);
    if (job->cfg.out_format==HTML) fputs("<br />",stdout);
    if (job->cfg.out_format!=XML)  fputc('\n', stdout);
    line = getTextLine(&(job->res.linelist), linecounter++);
  }
  free_textlines(&(job->res.linelist));
}

/* FIXME jb: remove JOB; renamed to OCR_JOB 2010-09-26 */
job_t *OCR_JOB;


/* -------------------------------------------------------------
// ------   MAIN - replace this by your own aplication! 
// ------------------------------------------------------------- */
int main(int argn, char *argv[]) {
  int multipnm=1;
  job_t job1, *job; /* fixme, dont want global variables for lib */
  job=OCR_JOB=&job1;

  setvbuf(stdout, (char *) NULL, _IONBF, 0);	/* not buffered */

  job_init(job); /* init cfg and db */

  process_arguments(job, argn, argv);
  
  /* load character data base (JS1002: now outside pgm2asc) */
  if ( job->cfg.mode & 2 ) /* check for db-option flag */
    load_db(job);
    /* load_db uses readpnm() and would conflict with multi images */
        
  while (multipnm==1) { /* multi-image loop */

    job_init_image(job); /* single image */

    mark_start(job);

    multipnm = read_picture(job);
    /* separation of main and rest for using as lib
       this will be changed later => introduction of set_option()
       for better communication to the engine  */
    if (multipnm<0) break; /* read error */
  
    /* call main loop */
    pgm2asc(job);

    mark_end(job);
  
    print_output(job);

    job_free_image(job);

  }

  return ((multipnm<0)?multipnm:0); /* -1=255 on error, 0 ok */
}
